File: src/kamishibai/input.py
"""Input document loading for kamishibai."""

import json
from typing import Protocol, final


class FieldMapping(Protocol):
    """Protocol for mapping JSON row to normalized entry dict."""

    def mapped(self, row):
        """Return normalized entry dict or None if row is invalid."""
        ...


@final
class VocabularyMapping:
    """Maps vocabulary JSON rows to normalized entry dicts."""

    def mapped(self, row):
        """Return normalized entry dict or None if row is invalid."""
        if not isinstance(row, dict):
            return None
        source = row.get("source")
        target = row.get("target")
        if not isinstance(source, dict) or not isinstance(target, dict):
            return None
        if not row.get("term") or not source.get("sentence") or not source.get("lang") or not target.get("sentence") or not target.get("lang"):
            return None
        return {
            "word": row["term"],
            "pronunciation": row.get("pronunciation") or "",
            "translation": row.get("meaning") or "",
            "example": target["sentence"],
            "source_lang": source["lang"],
            "target_lang": target["lang"],
            "sentence": source["sentence"],
            "highlight": source.get("highlight") or "",
            "hint": source.get("hint") or "",
            "context": source.get("context") or "",
            "importance": str(row.get("importance") or ""),
            "transcription": row.get("transcription") or "",
        }


@final
class Vocabulary:
    """Reads vocabulary entries from a JSON file."""

    def __init__(self, path, mapping):
        self._path = path
        self._mapping = mapping

    def document(self):
        """Load and validate the root JSON document."""
        with open(self._path, "r", encoding="utf-8") as file:
            data = json.load(file)
        if not isinstance(data, dict):
            raise ValueError(
                f"Expected a JSON object in '{self._path}' but found {type(data).__name__}"
            )
        if not isinstance(data.get("entries"), list):
            raise ValueError(
                f"Expected an 'entries' array in '{self._path}'"
            )
        return data

    def entries(self, document=None):
        """Load, filter, and return vocabulary entries."""
        data = self.document() if document is None else document
        result = []
        for row in data["entries"]:
            entry = self._mapping.mapped(row)
            if entry is not None:
                result.append(entry)
        if not result:
            raise ValueError(
                f"No valid entries found in '{self._path}'; each entry requires 'term', 'source.sentence', 'source.lang', 'target.sentence', and 'target.lang'"
            )
        return result

File: tests/input.rs
//! Tests for strict input parsing.

use std::fs;
use std::path::PathBuf;

use anyhow::Result;
use kamishibai::vocabulary::{
    Importance, LanguageCode, NonEmptyText, VocabularyDocument, VocabularyEntry, VocabularySource,
    VocabularyTarget,
};
use tempfile::TempDir;

/// Write one JSON document to a temporary file.
fn write(body: &str) -> Result<(TempDir, PathBuf)> {
    let directory = TempDir::new()?;
    let path = directory.path().join("kamishibai.json");
    fs::write(&path, body)?;
    Ok((directory, path))
}

/// Return one temporary path for one JSON string.
fn file(body: &str) -> Result<(TempDir, PathBuf)> {
    write(body)
}

/// Return one validated text fixture.
fn text(value: &str) -> NonEmptyText {
    NonEmptyText::new(value).expect("test text must be valid")
}

/// Return one validated language fixture.
fn code(value: &str) -> LanguageCode {
    LanguageCode::new(value).expect("test language must be valid")
}

/// Return one validated importance fixture.
fn score(value: u8) -> Importance {
    Importance::new(value).expect("test importance must be valid")
}

/// Valid entries keep the exact strict output shape.
#[test]
fn valid_entries_map_into_the_full_strict_shape() -> Result<()> {
    let (_directory, path) = file(
        r#"{"entries":[{"term":"кошка","meaning":"cat","pronunciation":"kæt","transcription":"kat","importance":7,"source":{"sentence":"Кошка спит на окне","lang":"ru","highlight":"Кошка","hint":"домашнее животное","context":"нейтральный стиль"},"target":{"sentence":"The cat is sleeping on the windowsill","lang":"en"}}]}"#,
    )?;
    assert_eq!(
        VocabularyDocument::load(&path)?,
        VocabularyDocument {
            entries: vec![VocabularyEntry {
                term: text("кошка"),
                meaning: text("cat"),
                pronunciation: text("kæt"),
                transcription: text("kat"),
                importance: score(7),
                source: VocabularySource {
                    sentence: text("Кошка спит на окне"),
                    lang: code("ru"),
                    highlight: text("Кошка"),
                    hint: text("домашнее животное"),
                    context: text("нейтральный стиль"),
                },
                target: VocabularyTarget {
                    sentence: text("The cat is sleeping on the windowsill"),
                    lang: code("en"),
                },
            }],
        },
        "valid entry did not map into the expected strict shape"
    );
    Ok(())
}

/// Missing required fields fail the whole document.
#[test]
fn missing_required_fields_fail_the_whole_document() -> Result<()> {
    let (_directory, path) = file(
        r#"{"entries":[{"term":"кошка","pronunciation":"kæt","transcription":"kat","importance":7,"source":{"sentence":"Кошка спит на окне","lang":"ru","highlight":"Кошка","hint":"домашнее животное","context":"нейтральный стиль"},"target":{"sentence":"The cat is sleeping on the windowsill","lang":"en"}}]}"#,
    )?;
    assert!(
        VocabularyDocument::load(&path)
            .unwrap_err()
            .to_string()
            .contains("missing field `meaning`"),
        "missing required fields no longer fail the whole document"
    );
    Ok(())
}

/// Empty strings fail the whole document.
#[test]
fn empty_strings_fail_the_whole_document() -> Result<()> {
    let (_directory, path) = file(
        r#"{"entries":[{"term":"кошка","meaning":"cat","pronunciation":" ","transcription":"kat","importance":7,"source":{"sentence":"Кошка спит на окне","lang":"ru","highlight":"Кошка","hint":"домашнее животное","context":"нейтральный стиль"},"target":{"sentence":"The cat is sleeping on the windowsill","lang":"en"}}]}"#,
    )?;
    assert!(
        VocabularyDocument::load(&path)
            .unwrap_err()
            .to_string()
            .contains("expected a non-empty string"),
        "empty strings no longer fail the whole document"
    );
    Ok(())
}

/// Unknown fields fail the whole document.
#[test]
fn unknown_fields_fail_the_whole_document() -> Result<()> {
    let (_directory, path) = file(
        r#"{"entries":[{"term":"кошка","meaning":"cat","pronunciation":"kæt","transcription":"kat","importance":7,"source":{"sentence":"Кошка спит на окне","lang":"ru","highlight":"Кошка","hint":"домашнее животное","context":"нейтральный стиль","tone":"soft"},"target":{"sentence":"The cat is sleeping on the windowsill","lang":"en"}}]}"#,
    )?;
    assert!(
        VocabularyDocument::load(&path)
            .unwrap_err()
            .to_string()
            .contains("unknown field `tone`"),
        "unknown fields no longer fail the whole document"
    );
    Ok(())
}

/// Invalid importance values fail the whole document.
#[test]
fn invalid_importance_values_fail_the_whole_document() -> Result<()> {
    let (_directory, path) = file(
        r#"{"entries":[{"term":"кошка","meaning":"cat","pronunciation":"kæt","transcription":"kat","importance":11,"source":{"sentence":"Кошка спит на окне","lang":"ru","highlight":"Кошка","hint":"домашнее животное","context":"нейтральный стиль"},"target":{"sentence":"The cat is sleeping on the windowsill","lang":"en"}}]}"#,
    )?;
    assert!(
        VocabularyDocument::load(&path)
            .unwrap_err()
            .to_string()
            .contains("expected an integer from 1 to 10"),
        "invalid importance values no longer fail the whole document"
    );
    Ok(())
}

/// Non-object roots fail with the frozen error wording.
#[test]
fn non_object_roots_raise_the_frozen_error_message() -> Result<()> {
    let (_directory, path) = file(r#"[{"term":"broken"}]"#)?;
    assert_eq!(
        VocabularyDocument::load(&path).unwrap_err().to_string(),
        format!(
            "Expected a JSON object in '{}' but found array",
            path.display()
        ),
        "non-object roots did not raise the frozen validation message"
    );
    Ok(())
}

/// Missing entries arrays fail with the strict error wording.
#[test]
fn missing_entries_arrays_raise_the_strict_error_message() -> Result<()> {
    let (_directory, path) = file(r#"{}"#)?;
    assert_eq!(
        VocabularyDocument::load(&path).unwrap_err().to_string(),
        format!(
            "Invalid document in '{}': missing field `entries`",
            path.display()
        ),
        "missing entries arrays did not raise the strict validation message"
    );
    Ok(())
}

/// Empty entry lists fail with the strict error wording.
#[test]
fn empty_entry_lists_raise_the_strict_error_message() -> Result<()> {
    let (_directory, path) = file(r#"{"entries":[]}"#)?;
    assert_eq!(
        VocabularyDocument::load(&path).unwrap_err().to_string(),
        format!("Expected at least one entry in '{}'", path.display()),
        "empty entry lists did not raise the strict validation message"
    );
    Ok(())
}
